crates/cubecl-opt/src/passes/find_safe_indexes.rs

```diff
@@ -0,0 +1,65 @@
+use cubecl_core::ir::{Operation, Operator, Variable, VariableKind};
+
+use crate::{
+    analyses::{const_len::Slices, integer_range::Ranges},
+    AtomicCounter, Optimizer,
+};
+
+use super::OptimizerPass;
+
+/// Use the results from integer range analysis to find indexes that are always in bounds, then
+/// transform them to unchecked indexes.
+pub struct InBoundsToUnchecked;
+
+impl OptimizerPass for InBoundsToUnchecked {
+    fn apply_post_ssa(&mut self, opt: &mut Optimizer, changes: AtomicCounter) {
+        let ranges = opt.analysis::<Ranges>();
```

Comment:
Yes it makes so much sense now! Optimization passes trigger analyses, which are cached!